### PR TITLE
shim: Start shims inside appropriate namespaces

### LINKED ISSUE #613: Shims related to the same container should be spawned into the same PID namespace
I was playing with docker recently, starting a container from a first shell:
```bash
docker run --runtime=runc --rm -it busybox
```
and exec'ing an extra process with the following command:
```bash
docker exec -it $CONTAINER_ID sh
```
when I realized that sometimes (I would say 50% of the time), the exec'ed process was not returning after an `exit` from the container process, and the shell was staying completely stuck until I restart my docker service.

I have investigated by looking at the difference between the logs from a working run vs a not working run, and I have found that when the shim corresponding to the container process was exiting before the shim related to the exec'ed process (basically a child process), we were having the wrong behavior described above.

The rationale behind the issue seems logical to me, because docker expects those processes to run inside a PID namespace, it does expect that the container process should be the last one to exit. Indeed, the container process being the "init" process of the PID namespace, when it terminates for any reason, the kernel SIGKILL all the other processes inside the PID namespace, and the "init" process returns.

Now, why are we getting issues in our case. The thing is that when our container process inside the VM terminates, the guest kernel will kill the processes inside the same PID namespace and they will all return `137` as exit code. But we are still inside the VM here, and we have to pass this information back to the shim (through a proxy in some cases), which means we have no control which shim (corresponding to every process) will exit first.

I have confirmed those investigations by getting this test working all the time by adding a 5s `sleep()` [here](https://github.com/clearcontainers/agent/blob/master/agent.go#L1127) in case the process was the container process. This way, I have forced the container process to wait 5s before to return its exit code back to the shim, leaving some times for other shims to return before the container process.

My suggestion for this issue is to solve this in 2 steps:
- The agent has to ensure every child process started with exec should be waited by the container process so that no process is left behind after the container process returns its exit code.
- Virtcontainers should start every shim related to the same container into the same PID namespace. This way, when the container process terminates, the host kernel can kill our shim processes related to exec'ed processes, all returning `137` exit code, or we can get the same exit code from the agent (in case the shim received the exit code before to be killed).

### PR DESCRIPTION
**Pull request by sboeuf**
This PR first creates a new package `nsenter` so that we can spawn any code into a set of namespaces. Then, the virtcontainers code is modified to rely on this new package. At last, it starts shims from agents within the right PID namespace. In details, a shim representing a container process should enter its own PID namespace so that we can spawn other shims corresponding to exec'ed processes inside the same namespace. This is needed if we don't want some race conditions to happen, due to the exec'ed shims to never return their exit codes if the container process returned first.

Fixes #613

### CODE REVIEW

File: pkg/nsenter/nsenter.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+// Namespace describes a namespace that will be entered.
+type Namespace struct {
+	Path string
+	PID  int
+	Type NSType
+}
+
+func getNSPathFromPID(pid int, nsType NSType) string {
+	return fmt.Sprintf("/proc/%d/ns/%s", pid, nsType)
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply pkg/nsenter/nsenter.go]: yes I will replace it.

File: pkg/nsenter/nsenter.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
+//
+// Copyright (c) 2018 Intel Corporation
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply pkg/nsenter/nsenter.go]: This is really a brand new re-implementation here. I have just got inspiration from what the `ns` package is doing for CNI.

File: pkg/nsenter/nsenter.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	nsFSMagic   = 0x6e736673
+	procFSMagic = 0x9fa0
+)
+
+// NSType defines a namespace type.
+type NSType string
+
+// List of namespace types.
+const (
+	NSTypeCGroup NSType = "cgroup"
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply pkg/nsenter/nsenter.go]: ;)

File: pkg/nsenter/nsenter.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+	file, err := os.Open(nsPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return file, nil
+}
+
+func sortNSList(nsList []Namespace) []Namespace {
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply pkg/nsenter/nsenter.go]: Definitely !

File: pkg/nsenter/nsenter.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+		}
+	}
+
+	// No PID ns found, return the unmodified list.
+	return nsList
+}
+
+// NsEnter executes the passed closure under the given namespace,
+// restoring the original namespace afterwards.
+func NsEnter(unsortedNSList []Namespace, toRun func() error) error {
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply pkg/nsenter/nsenter.go]: Yes, I was planning on implementing some testing here. I'll work on that today.

**Comment by amshinde** [comment]
@sboeuf Did some digging up on the issues filed surrounding the mount namespace issue when I was implementing support for devices. See this:
https://github.com/golang/go/issues/8676

Also, that issue has code that checks if the namespace change was successful, we could do something similar in our tests:
https://storage.googleapis.com/go-attachment/8676/0/checkns.c

**Comment by amshinde** [comment]
@sboeuf Doing things the way you are doing currently works (most of the time) and provides for a simple solution (instead of reexecing a new process), but I came across this article which does state that it is not 100% foolproof : https://www.weave.works/blog/linux-namespaces-and-go-don-t-mix

Basically golang may spawn a new thread for a blocking syscall even after LockOsThread()
We have an issue in vishwanda/netlink repo related to the above:
https://github.com/vishvananda/netns/issues/17

The issue seems to have been resolved in golang 1.10, but we may see an unsuccessful setns call in earlier versions.

Doing things in the shim code, may not be such a bad approach after all.

**Comment by sboeuf** [comment]
@jodh-intel PR updated with almost full coverage for unit tests and more than that a real test that we're entering namespaces.
As @amshinde explained, we have some limitations with Go, even if this code is supposed to be foolproof.
- About limitations, both mount and user namespaces cannot be entered with the `Setns()` syscall from a Go program because it's multithreaded. I honestly don't get the "why" of this limitation but after all I have read and my local testing, I can confirm we cannot enter those namespaces if we don't involve some C code, or the use of an external binary such as `nsenter`.
- About being foolproof, I think this is safe since we know that we should not pass a callback involving a go routine through this package, otherwise it won't be executed in the expected namespace. Also, the second issue about having some new threads being created in case the current one takes too long should not happen in our case and will be solved with Go 1.10.

Basically, this package does the most that a Go package can do with netns. So my proposal is that we could take it in since it will help spawning our shims into expected PID and network namespaces, and worst case, it won't do any harm in case we hit the issues explained above once every 10000000000 times.

At least this whole PR brings the structure for the shims to be switched to the expected namespaces.

I think we should try to look into a better way to solve this, but this will involve more researches.

File: pkg/nsenter/nsenter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sameo) [reply pkg/nsenter/nsenter.go]: If there's *anything* that was loosely copied from the ns package, we have to include the right copyrights.

[the remark this answers is not in the document]
Comment (sboeuf) [reply pkg/nsenter/nsenter.go]: Ok I understand, then I'll add them.

**Comment by sboeuf** [comment]
@sameo @jodh-intel @amshinde PR is ready for another review ;)

File: pkg/nsenter/nsenter_test.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+			unix.Kill(sleepPIDFromNsEnter, syscall.SIGKILL)
+		}
+	}()
+
+	for _, ns := range nsList {
+		nsPathEntered := getNSPathFromPID(sleepPIDFromNsEnter, ns.Type)
+
+		evalExpectedNSPath, err := filepath.EvalSymlinks(ns.Path)
+		if err != nil {
+			evalExpectedNSPath = err.(*os.PathError).Path
```
Comment (amshinde) [inline pkg/nsenter/nsenter_test.go]: Might be useful to add comments here describing how you are doing the comparison.

File: hyperstart_agent.go [hunk 7]
```diff
@@ -314,7 +315,22 @@ func (h *hyper) exec(pod *Pod, c Container, cmd Cmd) (*Process, error) {
 		Process:   *hyperProcess,
 	}
 
-	process, err := prepareAndStartShim(pod, h.shim, c.id, token, h.state.URL, cmd)
```
Comment (amshinde) [inline hyperstart_agent.go]: I prefer passing additional arguments here rather than duplicating the code.

File: pkg/nsenter/nsenter.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+	}
+
+	if err := unix.Setns(int(nsFile.Fd()), nsFlag); err != nil {
+		return fmt.Errorf("Error switching to ns %v: %v", nsFile.Name(), err)
+	}
+
+	return nil
+}
+
+func getFileFromNS(nsPath string) (*os.File, error) {
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply pkg/nsenter/nsenter.go]: done

File: pkg/nsenter/nsenter.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	var innerError error
+	go func() {
+		defer wg.Done()
+		runtime.LockOSThread()
+		innerError = containedCall()
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply pkg/nsenter/nsenter.go]: Yes sure I think we can do that. But it won't change anything since this goroutine ends after that.

File: pkg/nsenter/nsenter_test.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+	assert.NotNil(t, err, "Should fail because neither a path nor a PID"+
+		" has been provided by every namespace of the list")
+}
+
+func TestNsEnterEmptyNamespaceListSuccess(t *testing.T) {
+	err := NsEnter([]Namespace{}, testToRunNil)
+	assert.Nil(t, err, "Should not fail since closure should return nil: %v", err)
+}
+
+func areNSIdentical(nsPath1, nsPath2 string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (sboeuf) [reply pkg/nsenter/nsenter_test.go]: Oh yeah you're right, that was a temp function that I should have removed !

File: pkg/nsenter/nsenter_test.go (hunk 6 above)
Comment (sboeuf) [reply pkg/nsenter/nsenter_test.go]: Yes good point

File: hyperstart_agent.go (hunk 7 above)
Comment (sboeuf) [reply hyperstart_agent.go]: Fair enough ! I have removed it.

**Comment by sboeuf** [comment]
@amshinde PR updated, PTAL

**Comment by sboeuf** [comment]
@grahamwhaley @jodh-intel  ready to be merged
